File: algorithmic_toolbox/week2_algorithmic_warmup/1_fibonacci_number/Fibonacci.java
import java.util.Scanner;

public class Fibonacci {
  public static long calcFibNaive(int n) {
    return n <= 1 ? n : calcFibNaive(n - 1) + calcFibNaive(n - 2);
  }

  public static long calcFibFast(int n) {
    if (n <= 1)
      return n;

    long[] fibNums = new long[n + 1];
    fibNums[0] = 0;
    fibNums[1] = 1;

    for (int i = 2; i <= n; i++)
      fibNums[i] = fibNums[i - 1] + fibNums[i - 2];

    return fibNums[n];
  }

  public static void main(String args[]) {
    Scanner scanner = new Scanner(System.in);
    int n = scanner.nextInt();
    scanner.close();

    System.out.println(calcFibFast(n));
  }
}

File: algorithmic_toolbox/week2_algorithmic_warmup/1_fibonacci_number/fibonacci.py
def calc_fib_naive(n: int) -> int:
    return n if n <= 1 else calc_fib_naive(n - 1) + calc_fib_naive(n - 2)


def calc_fib_fast(n: int) -> int:
    if n <= 1:
        return n

    fib_nums: list[int] = [0, 1]

    for i in range(2, n + 1):
        fib_nums.append(fib_nums[i - 1] + fib_nums[i - 2])

    return fib_nums[n]


if __name__ == "__main__":
    n = int(input())
    print(calc_fib_fast(n))
